Learning path(fast): Add adaptive serving guide and codelab. Fixes #765, Fixes #665.

robdodson opened the pull request:
Fixes #765, #665.

Changes proposed in this pull request:

- Add adaptive serving guide.
- Add adaptive serving codelab.

Staged at: https://20190501t152948-dot-web-dev-staging.appspot.com/adaptive-serving-based-on-network-quality/

robdodson commented:
A few things worth noting:
- I removed smart quotes and replaced them with single quotes. I've had folks accidentally copy/pate smart quotes during a codelab once and they couldn't figure out why their JS wouldn't compile because their editor styled them similarly.
- The link for codelabs appears at the bottom now. You can also add an inline callout for someone to go checkout your codelab by doing this in your markdown:
```
{% Aside 'codelab' %}
Check out [this awesome codelab](link/to/codelab)
{% endAside %}
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+To load it conditionally, first check if connection type detection is supported.
+In `script.js` add an if statement that tests whether the `navigator.connection`
+object exists and whether it has the `effectiveType` property.
```
(Recommend putting `if` and `else` in code font.)

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
````diff
@@ -0,0 +1,210 @@
+}
+```
+
+Wrap the existing video loading code in an else statement, so that video will
````
```suggestion
Wrap the existing video loading code in an `else` statement, so that video will
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+## Step 2: Load video
+
+If the `effectiveType` is `'4g'`, you can use the video loading code from the
```
```suggestion
If the `effectiveType` is `'4g'`, use the video loading code from the
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+## Step 1: Check connection type
+
+The background video is currently specified on line 22 of `index.html` and then
```
Also: since there's a longer explanation of this later on, maybe have a reference here, like, "The video loading code is described in more detail in [section]"

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
````diff
@@ -0,0 +1,210 @@
+}
+```
+
+Here’s how the video loading code works: the video source is specified inside
````
```suggestion
Here’s how the video loading code works: the video source is specified using the
```

arthurevans replied:
Alternately, maybe explain a little more here. Like:

.. code works: the `<video>` tag doesn't download or display anything at first because its `src` attribute is not set. The video URL to load is specified using the `data-src` attribute.

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+Here’s how the video loading code works: the video source is specified inside
+[data-src attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
+in `index.html` and then lazy-loaded in `script.js`.
```
```suggestion
on the `<video>` tag.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+Here’s how the video loading code works: the video source is specified inside
+[data-src attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
````
Don't need two links to this doc here. Recommend maybe cutting this one and only including the following one. Also: `data-src` should be in code font.

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+Because the video source is specified in the `data-src` attribute, it isn’t
+displayed or downloaded initially.
+[Data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
+allow you to store extra information on standard HTML elements. They can be
```
```suggestion
allow you to store extra information on standard HTML elements. A data element can be
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+named anything, as long as it starts with "data-".
+
+To actually display the video on the page, the location from `data-src` needs to
+be set as the `src` attribute of the video element. This is done in `script.js`.
```
"the location from ..." maybe rephrase as "you need to copy the value from `data-src` to the video element's `src` attribute."

I think . you can leave off the "This is done in script.js."

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+To test it yourself:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
```
"press the **App** button  ... " can be shortened to "click **App**, then click **Show** to preview the app."

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+1. Mouse over the editor, press the **App** button, then press the **Show**
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
```
```suggestion
3. Click the **Network** tab.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+<img class="w-screenshot" src="./devtools_network_throttling.png" alt='DevTools Network tab with Fast 3G throttling option highlighted'>
+
+Now reload the page with Fast 3G still enabled and the app will load an image in the background instead of the video:
```
```suggestion
Now reload the page with Fast 3G still enabled. The app loads an image in the background instead of the video:
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+## Extra Credit: Respond to changes
+
+Remember how this API has an `onchange`
+[event listener](http://localhost:8080/fast/adaptive-serving-based-on-network-quality#how-it-works)?
```
Bad URL. (I think just cut http://localhost:8080 and make it an absolute path?)

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+Remember how this API has an `onchange`
+[event listener](http://localhost:8080/fast/adaptive-serving-based-on-network-quality#how-it-works)?
+You can use it for many things: dynamically adapt content such as video quality, restart deferred data transfers when a change to a high-bandwidth network type is detected, or notify users when the network quality changes.
```
```suggestion
You can use it for many things: dynamically adapting content such as video quality; restarting deferred data transfers when a change to a high-bandwidth network type is detected; or notifying users when the network quality changes.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+[event listener](http://localhost:8080/fast/adaptive-serving-based-on-network-quality#how-it-works)?
+You can use it for many things: dynamically adapt content such as video quality, restart deferred data transfers when a change to a high-bandwidth network type is detected, or notify users when the network quality changes.
+
+Here’s a simple example of how this listener works, add it to `script.js`. This
```
```suggestion
Here’s a simple example of how to use this listener. Add it to `script.js`. This
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+
+To test it again:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
```
```suggestion
1. Mouse over the editor, click **App** button, then click **Show**
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+To test it again:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
+   button to preview the app.
```
```suggestion
   to preview the app.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+1. Mouse over the editor, press the **App** button, then press the **Show**
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
```
```suggestion
3. Click the **Network** tab.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
+4. Select **Fast 3G** from throttling presets.
```
```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
+4. Select **Fast 3G** from throttling presets.
```
```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

arthurevans commented on src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,210 @@
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
+4. Select **Fast 3G** from throttling presets.
+5. **Reload** the page.
```
```suggestion
5. Reload the page.
```

arthurevans commented on src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,125 @@
+* Defer uploads and downloads when users are on a slow connection. 
+* Enable offline mode if the network quality is not good enough to load the app
+  and use the features.
+* Warn users that doing something (e.g. watching video) over cellular could cost
```
```suggestion
* Warn users that doing something (for example, watching video) over cellular could cost
```

robdodson commented on src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,125 @@
+Loading a website can be a very different experience depending on the network
+conditions. Everything is usually smooth when you are on a fast network, but
+when you're on the go with a humble data plan and spotty connection, or stuck
+with a laptop on a sluggish coffee shop wifi, it's a different story.
```
Added `Wi-Fi` to our guide here: https://github.com/GoogleChrome/web.dev/wiki/Writing-style#spelling-and-capitalisation

mihajlija commented:
@arthurevans Thank you for the thoughtful review! 😊 I've fixed the formatting and updated the text, you can take another look when you get a chance. 

I don't have the access privileges to build the site locally, so please double-check the links in the codelab which reference the guide (especially the one with the colon, on line 25).

robdodson commented:
Hey Milica, sorry about that. I just added you to the private repo that
will let you pull in the styles.
https://github.com/GoogleChrome/web.dev-frontend/invitations

*From: *Milica <notifications@github.com>
*Date: *Sun, May 5, 2019 at 4:21 PM
*To: *GoogleChrome/web.dev
*Cc: *Rob Dodson, Author

@arthurevans <https://github.com/arthurevans> Thank you for the thoughtful
> review! 😊 I've fixed the formatting and updated the text, you can take
> another look when you get a chance.
>
> I don't have the access privileges to build the site locally, so please
> double-check the links in the codelab which reference the guide (especially
> the one with the colon, on line 25).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleChrome/web.dev/pull/857#issuecomment-489472403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIEKDJYAYMDOUQ6PO35IFTPT5TVTANCNFSM4HJZTATQ>
> .
>

mihajlija commented:
@robdodson Thanks! I've checked the preview, the links in the content work, but returning from a codelab to the associated guide seems to be broken. Looks like the return button points to the old URL scheme. Please ignore this if it's just a preview issue :)

![Screen Shot 2019-05-06 at 11 38 55 AM](https://user-images.githubusercontent.com/17127567/57217913-3695c880-6ff4-11e9-86c1-c1bda060ec07.png)

![Screen Shot 2019-05-06 at 11 44 22 AM](https://user-images.githubusercontent.com/17127567/57217992-73fa5600-6ff4-11e9-8291-85f6da646ad5.png)

robdodson commented:
@mihajlija ah I think the return button is fixed in current master.

If you checkout the master branch, do a git pull, then checkout your feature branch and run `git rebase master` it should fix it.

Taking a quick look at the post now to see if we can land it.

robdodson commented:
Looks good to me, let's ship it! 🚢
